[quant][graphmode] Test refactor -- extract graph checker

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36994 [quant][graphmode] Test refactor -- extract graph checker**
* #36889 [quant][graphmode] linear_relu

Differential Revision: [D21225873](https://our.internmc.facebook.com/intern/diff/D21225873)

Commit:
[quant][graphmode] Test refactor -- extract graph checker

[ghstack-poisoned]

z-a-f commented:
@jerryzh168 This is a proposal refactor -- if agreed, I will replace all `_test_op_impl` with the factored out one

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1191,13 +1192,12 @@ def forward(self, x):
-                   .run(model.graph)
+        model, checker = test_module_graph(
+            M(), data,
+            checks=[['check', ("quantized::linear",)],
+                    ['check_not', ("quantized::linear_prepack",)],
+                    ['check_count', ("aten::quantize_per_tensor", 1, True)]
+                    ])
```
I think maybe we can just have three lists, one for check, one for check_not and one for check_count

jerryzh168 replied:
I think you can do two lists, if user want to do check_count they can do it outside of the function.

z-a-f replied:
In that case we cannot have an ordered check. Every check will be a separate run.

jerryzh168 replied:
that's OK, I think the purpose of the helper function is to simplify the common cases, we don't have to cover all cases. we can return the model can people can run further checks on them

jerryzh168 replied:
when we cover all cases it becomes too complex and people might as well just call the original checker..

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,77 @@
+"""
+def test_module_graph(module, data=None, checks=None, ordered_checks=None,
+                      qengine=None):
+    def _make_check_type(check_type):
+        if check_type[:5] == 'check':
+            return check_type
+        if check_type != '':
+            check_type = '_' + check_type
+        check_type = 'check' + check_type
+        return check_type
```
so that we don't need to parse the check type here

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,77 @@
+    checker: `FileCheck` instance
+"""
+def test_module_graph(module, data=None, checks=None, ordered_checks=None,
+                      qengine=None):
```
should we default qengine to fbgemm?

Commit:
Update on "[quant][graphmode] Test refactor -- extract graph checker"




[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] Test refactor -- extract graph checker"


Differential Revision: [D21225873](https://our.internmc.facebook.com/intern/diff/D21225873)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] Test refactor -- extract graph checker"


Differential Revision: [D21225873](https://our.internmc.facebook.com/intern/diff/D21225873)

[ghstack-poisoned]

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,113 @@
+                                     ('check_not', 'aten::linear')))
+"""
+def _test_module_graph(module, data=None, check=None, check_not=None,
+                      check_count=None, ordered_checks=None, qengine=None):
```
I think we can remove check_count and ordered_checks as well

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,113 @@
+    assert qengine in ['fbgemm', 'qnnpack'], \
+        'Qengine must be one of (fbgemm, qnnpack)'
+
+    if isinstance(module, torch._C.Graph):
+        # If already a graph, just run the checks
+        model = None
+        model_graph = module
+        model_graph_count = module
+    elif isinstance(module, torch.jit.ScriptModule):
+        # Check if the module is already scripted
+        model = module
+    else:
+        qconfig_dict = {'': get_default_qconfig(qengine)}
+        model = torch.jit.script(module).eval()
+        model = quantize_script(model, qconfig_dict, test_only_eval_fn, [data],
+                                inplace=False)
```
also, let's just assume ScriptModule here. we don't need to do checks

jerryzh168 replied:
oh, sorry, let's always quantize here?

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,113 @@
+        model = quantize_script(model, qconfig_dict, test_only_eval_fn, [data],
+                                inplace=False)
+
+    # Check if data exists
+    if data is None:
+        if model is not None:
+            model_graph = model.graph
+    else:
+        *inputs, target = data[0]
+        # Make sure it runs
+        model(*inputs)
+        # We could use `graph_for(*input)` but it might break the `check_count`.
+        model_graph = model.graph_for(*inputs)
+        model_graph_count = model.graph
```
we can just use model.graph here, we call finalize by default so the graphs will be inlined by default now, and there are graphs that's not runnable in the tests

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,113 @@
+            check_not = (check_not,)
+        for c in check_not:
+            FileCheck().check_not(c).run(model_graph)
+    if check_count is not None:
+        if not isinstance(check_count, (list, tuple)):
+            raise ValueError("`check_count` must be a list of lists")
+        if len(check_count) == 3 and isinstance(check_count[0], str):
+            check_count = (check_count,)  # tuple of iterables
+        for c in check_count:
+            FileCheck().check_count(*c).run(model_graph_count)
+    if ordered_checks is not None:
+        checker = FileCheck()
+        for check_type, check_args in ordered_checks:
+            if not isinstance(check_args, (list, tuple)):
+                check_args = (check_args,)
+            getattr(checker, check_type)(*check_args)
+        checker.run(model_graph)
```
these can be removed, we can call these outside of this function.

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -12,12 +12,113 @@
+
+    # Run the checks
+    if check is not None:
+        if not isinstance(check, (list, tuple)):
```
can we just always pass in a list/tuple?

jerryzh168 reviewed (changes requested):
In general, `_test_module_graph` is just an internal helper function, I think we can assume many things and make the API more restrictive to reduce the code I write for this function.

jerryzh168 commented:
Also, please fix the python lint errorrs
